File: packages/android/app/src/main/kotlin/dev/voile/lib/LogError.kt
package dev.voile.lib

import io.sentry.Sentry
import io.sentry.SentryLevel

/**
 * Log une erreur métier avec contexte Voile.
 */
fun logError(
    throwable: Throwable,
    feature: String? = null,
    action: String? = null,
    serverId: Int? = null,
    extra: Map<String, Any>? = null,
) {
    // Console
    System.err.println("[Voile] ${feature ?: "unknown"}.${action ?: "unknown"}: $throwable")
    throwable.printStackTrace()

    // Sentry
    Sentry.withScope { scope ->
        feature?.let { scope.setTag("feature", it) }
        action?.let { scope.setTag("action", it) }
        serverId?.let { scope.setTag("server_id", it.toString()) }

        if (extra != null) {
            scope.setExtras(extra)
        }

        scope.level = SentryLevel.ERROR
        Sentry.captureException(throwable)
    }
}

/**
 * Log un breadcrumb (info).
 */
fun logInfo(message: String, data: Map<String, Any>? = null) {
    Sentry.addBreadcrumb(
        io.sentry.Breadcrumb().apply {
            this.message = message
            this.level = SentryLevel.INFO
            if (data != null) this.data = data
        }
    )
}

/**
 * Démarre une transaction de performance.
 */
inline fun <T> trackPerformance(
    name: String,
    operation: String = "function",
    block: () -> T,
): T {
    return Sentry.startTransaction(name, operation).let { transaction ->
        try {
            block()
        } finally {
            transaction.finish()
        }
    }
}

File: packages/web/src/lib/logError.ts
import * as Sentry from "@sentry/react";

interface ErrorContext {
  feature?: string;
  action?: string;
  server?: { id?: number; country?: string; city?: string };
  extra?: Record<string, unknown>;
}

/**
 * Log une erreur métier avec contexte Voile.
 * Envoie à Sentry si configuré, console.warn sinon (dev).
 */
export function logError(error: unknown, context: ErrorContext = {}): void {
  // Log console toujours (debug + sentry ingestion logs)
  console.error(`[Voile] ${context.feature ?? "unknown"}.${context.action ?? "unknown"}:`, error);

  // Sentry
  Sentry.withScope((scope) => {
    if (context.feature) scope.setTag("feature", context.feature);
    if (context.action) scope.setTag("action", context.action);
    if (context.server?.id) {
      scope.setContext("server", {
        id: context.server.id,
        country: context.server.country ?? "unknown",
        city: context.server.city ?? "unknown",
      });
    }
    if (context.extra) scope.setExtras(context.extra);
    scope.setLevel("error");
    Sentry.captureException(error);
  });
}

/**
 * Log un message informatif (non-erreur).
 */
export function logInfo(message: string, data?: Record<string, unknown>): void {
  Sentry.addBreadcrumb({
    message,
    data,
    level: "info",
  });
}

/**
 * Démarre une transaction de performance.
 */
export function startTransaction(name: string, op: string) {
  return Sentry.startSpan({ name, op }, () => {});
}
